File: server/game-server/app/game/trigger.ts
import { GameActionBase } from "./action/gameActionBase";
import { ACTION_TYPE } from "./constants";
/**
 * 触发优先级
 */
export enum TRIGGER_PRIORITY {
	LOW,
	NORMAL,
	HIGH,
}
export type TriggerId = number;
export default class Trigger {
	private _triggerMap: Map <ACTION_TYPE, Set<TriggerId> >;
	private _tid: TriggerId = 0;
	private _tidMap: Map< TriggerId, (args: GameActionBase) => void >;
	constructor () {
		this._tidMap = new Map();
		this._triggerMap = new Map();

	}
	public register (type: ACTION_TYPE, effect: (args: GameActionBase) => void): TriggerId {
		if (!this._triggerMap.has(type)) {
			this._triggerMap.set(type, new Set());
		}
		this._tid++;
		this._triggerMap.get(type).add(this._tid);
		this._tidMap.set(this._tid, effect);
		return this._tid;
	}
	public notify (action: GameActionBase) {
		const type = action.type;
		if (!this._triggerMap.has(type)) { return; }
		const tids = this._triggerMap.get(type);
		if (!tids) { return; }
		for (const tid of tids) {
			if (!this._tidMap.has(tid)) {
				tids.delete(tid);
			} else {
				this._tidMap.get(tid)(action);
			}
		}
	}
	public remove (tid: TriggerId) {
		if (this._tidMap.has(tid)) {
			this._tidMap.delete(tid);
		}
	}
}

File: server/Logic/Game/Action/Trigger.cs
using System;
using System.Collections.Generic;
namespace Logic.Action
{

	class Trigger
	{
		private Dictionary<ACTION_TYPE, List<int>> _triggerMap;//保证触发顺序，这里用list
		private int _tid = 0;
		private Dictionary<int, Action<GameActionBase>> _tidMap;
		public Trigger()
		{
			this._tidMap = new Dictionary<int,Action<GameActionBase>>();
			this._triggerMap = new Dictionary<ACTION_TYPE, List<int>>();
		}
		public int register(ACTION_TYPE type, Action<GameActionBase> effect)
		{
			if (!this._triggerMap.ContainsKey(type))
			{
				this._triggerMap.Add(type, new List<int>());
			}
			this._tid++;
			this._triggerMap[type].Add(this._tid);
			this._tidMap.TryAdd(this._tid, effect);
			return this._tid;
		}
		internal void notify(GameActionBase action)
		{
			ACTION_TYPE type = action.type;
			if (!this._triggerMap.ContainsKey(type)) { return; }
			List<int> tidList = this._triggerMap[type];
			if (tidList.Count == 0) { return; }
			HashSet<int> toDelete = new HashSet<int>();
			foreach (int tid in tidList)
			{
				if (!this._tidMap.ContainsKey(tid))
				{
					toDelete.Add(tid);
				}
				else
				{
					this._tidMap[tid](action);
				}
			}
			foreach (int tid in toDelete)
			{
				tidList.Remove(tid);
			}
		}
		public void remove(int tid)
		{
			this._tidMap.Remove(tid);
		}
	}

}